### PR TITLE
fix #309, add register order for router package

### LINKED ISSUE #309: need more powerful extension support
if we need to extend some protocols or override it like  sofarpc

we need to provide more powerful extension support ，for extensions ，they can specify their orders，extension manager will use  the extension with biggest order

### PR DESCRIPTION
**Pull request by nejisama**
### Issues associated with this PR

#309 
### Sign the CLA
Make sure you have signed the [CLA](https://www.clahub.com/agreements/alipay/sofa-mosn)

### Solutions

Register with order, if new register order is bigger than older order, register success, or ignore the register

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases.
And you need to show the ut result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/router/factory.go [hunk 1]
```diff
@@ -21,18 +21,23 @@ import (
...
+var defaultRouterRuleFactoryOrder routerRuleFactoryOrder
 
-func RegisterRouterRule(f RouterRuleFactory) {
-	defaultRouterRuleFactory = f
+func RegisterRouterRule(f RouterRuleFactory, order uint32) {
+	if defaultRouterRuleFactoryOrder.order < order {
+		log.DefaultLogger.Infof("register router rule, order %d", order)
+		defaultRouterRuleFactoryOrder.factory = f
+		defaultRouterRuleFactoryOrder.order = order
+	}
```
Comment (ldxdl) [inline pkg/router/factory.go]: 这里如果因为order没有注册成功最后有一行warning

File: pkg/router/factory.go [hunk 2]
```diff
@@ -47,10 +52,14 @@ func DefaultSofaRouterRuleFactory(base *RouteRuleImplBase, headers []v2.HeaderMa
...
+var makeHandlerChainOrder handlerChainOrder
 
-func RegisterMakeHandlerChain(f MakeHandlerChain) {
-	makeHandlerChain = f
+func RegisterMakeHandlerChain(f MakeHandlerChain, order uint32) {
+	if makeHandlerChainOrder.order < order {
+		log.DefaultLogger.Infof("register make handler chain, order %d", order)
+		makeHandlerChainOrder.makeHandlerChain = f
+		makeHandlerChainOrder.order = order
+	}
```
Comment (ldxdl) [inline pkg/router/factory.go]: 这里如果因为order没有注册成功最后有一行warning

File: pkg/router/types.go [hunk 3]
```diff
@@ -203,3 +203,13 @@ type RouterRuleFactory func(base *RouteRuleImplBase, header []v2.HeaderMatcher)
 
 // MakeHandlerChain creates a RouteHandlerChain
 type MakeHandlerChain func(types.HeaderMap, types.Routers, types.ClusterManager) *RouteHandlerChain
+
+// reigister order
+type routerRuleFactoryOrder struct {
```
Comment (ldxdl) [inline pkg/router/types.go]: 缺少注释 最起码说明一下order的作用，有效范围，生效方式
